Fix edge condition where we try to connect an invalid chain but need to finish syncing

ptschip opened the pull request:
This can under some circumstances prevent us from syncing fully
to the correct chain until we receive the next block announcement.

This was happening in the ctor.py test where we had two chains, one
which was DTOR and one CTOR and then we tried to connect a node
to these two chains and let it sync.  Sometimes it was hanging
because when one chain was marked invalid then we ended up
setting the chaintip to the most work chain. This casues us to
connect all the blocks in the setblockindexcandidates resulting
in the pindexBestHeader being equal to the chainTip.  The final result
is that no more blocks could be downloaded because it appeared that
the chain was fully synced. Another new block announcement would be
needed to get the sync to continue.  This problem is unlikely to
be seen on mainnet, however it could happen with the end result that
our chain sync would fall behind until the next block was mined.

ptschip committed:
Fix edge condition where we try to connect an invalid chain

This can under some circumstances prevent us from syncing fully
to the correct chain until we receive the next block announcement.

This was happening in the ctor.py test where we had two chains, one
which was DTOR and one CTOR and then we tried to connect a node
to these two chains and let it sync.  Sometimes it was hanging
because when one chain was marked invalid then we ended up
setting the chaintip to the most work chain. This casues us to
connect all the blocks in the setblockindexcandidates resulting
in the pindexBestHeader being equal to the chainTip.  The final result
is that no more blocks could be downloaded because it appeared that
the chain was fully synced. Another new block announcement would be
needed to get the sync to continue.  This problem is unlikely to
be seen on mainnet, however it could happen with the end result that
our chain sync would fall behind until the next block was mined.

gandrewstone commented on src/net_processing.cpp:
```diff
@@ -2517,7 +2517,15 @@ bool SendMessages(CNode *pto)
+        // best invalid chain work is still greater than our chaintip then we have to keep looking for more blocks
+        // to download.
+        if (!IsChainSyncd() || (pindexBestInvalid.load() && chainActive.Tip() &&
+                                   pindexBestInvalid.load()->nChainWork > chainActive.Tip()->nChainWork))
```
you need to store pindexBestInvalid.load() and chainActive.Tip() in temporaries since they could be nullified between the 2 calls.

ptschip replied:
fixed

gandrewstone commented on src/validation/validation.cpp:
```diff
@@ -425,7 +425,7 @@ bool LoadBlockIndexDB()
             setBlockIndexCandidates.insert(pindex);
         if (pindex->nStatus & BLOCK_FAILED_MASK &&
-            (!pindexBestInvalid || pindex->nChainWork > pindexBestInvalid->nChainWork))
+            (!pindexBestInvalid || pindex->nChainWork > pindexBestInvalid.load()->nChainWork))
```
store in temporary

ptschip replied:
fixed

ptschip committed:
Use temporaries when evaluating atomics in an 'if' statment

In a couple of places we get the atomic value twice in a if statment
but it could have been nullified between the calls.
